src/persist/benches/consensus.rs

```diff
@@ -0,0 +1,95 @@
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+//! Benchmarks for different persistent Write implementations.
+
+use std::path::Path;
+use std::time::{Duration, Instant};
+
+use criterion::measurement::WallTime;
+use criterion::{Bencher, BenchmarkGroup, BenchmarkId, Throughput};
+
+use mz_persist::location::{Consensus, SeqNo, VersionedData};
+use mz_persist::mem::MemConsensus;
+use mz_persist::sqlite::SqliteConsensus;
+use mz_persist::workload::{self, DataGenerator};
+
+fn new_sqlite_consensus(path: &Path) -> SqliteConsensus {
+    SqliteConsensus::open(&path).expect("creating a SqliteConsensus cannot fail")
+}
+
+// Benchmark the write throughput of Consensus::compare_and_set.
+fn bench_compare_and_set<C: Consensus>(
+    consensus: &mut C,
+    key: &str,
+    data: Vec<u8>,
+    b: &mut Bencher,
+) {
+    let deadline = Instant::now() + Duration::from_secs(3600);
+
+    // Set an initial value in case that has different performance characteristics
+    // for some Consensus implementations.
+    //
+    // The benchmark runner might execute this code several times to warm-up, grab
+    // the existing value if one has already been set.
+    let mut current = match futures_executor::block_on(consensus.compare_and_set(
+        key,
+        deadline,
+        None,
+        VersionedData {
+            seqno: SeqNo(0),
+            data: data.clone(),
+        },
+    ))
+    .expect("gave invalid inputs")
+    {
+        Ok(()) => SeqNo(0),
+        Err(Some(versioned_data)) => versioned_data.seqno,
+        // We know the return value must be Some(_) because otherwise, the
+        // compare_and_set call should have succeeded with expected=None.
+        Err(None) => unreachable!(),
+    };
+    b.iter(|| {
+        let next = current.next();
+        futures_executor::block_on(consensus.compare_and_set(
+            key,
+            deadline,
+            Some(current),
+            VersionedData {
+                seqno: next,
+                data: data.clone(),
+            },
+        ))
+        .expect("gave invalid inputs")
+        .expect("failed to compare_and_set");
+        current = next;
+    })
+}
+
+pub fn bench_consensus_compare_and_set(data: &DataGenerator, g: &mut BenchmarkGroup<'_, WallTime>) {
+    let blob_val = workload::flat_blob(&data);
+    g.throughput(Throughput::Bytes(data.goodput_bytes()));
+
+    let mut mem_consensus = MemConsensus::default();
+    g.bench_with_input(
+        BenchmarkId::new("mem", data.goodput_pretty()),
+        &blob_val,
+        |b, blob_val| bench_compare_and_set(&mut mem_consensus, "mem", blob_val.clone(), b),
+    );
+
+    // Create a directory that will automatically be dropped after the test finishes.
+    let temp_dir = tempfile::tempdir().expect("failed to create temp directory");
+    let consensus_path = temp_dir.path().join("bench_compare_and_set");
```

Comment:
shouldn't this be something like `join("sqlite")` in case we use this temp_dir for benchmarking other impls?

also super nit: inline this consensus_path bit in the new_sqlite_consensus call on the line below, please. or call it sqlite_path